cli,ci,ipsec: create key command and CI integration

viktor-kurchenko opened the pull request:
Please see per commit description.

viktor-kurchenko committed:
cli,ipsec: create key command

The commit adds IPsec key create sub-command in Cilium CLI.
The intention is to use this command for encryption secret provisioning
before the Cilium installation.

Signed-off-by: viktor-kurchenko <viktor.kurchenko@isovalent.com>

smagnani96 reviewed (commented):
I remember discussing (and tried to implement) this with @pchaigno during my on-boarding. IIRC, the outcome of the discussion was "It's a nice to have, but the burden of maintaining additional code might be heavier than the advantages brought".

Though, happy to re-consider this :smiley: 
The code LGTM. We're not going to support all possible cases (ex looking [here](https://github.com/cilium/cilium/blob/main/pkg/datapath/linux/ipsec/ipsec_linux_test.go#L51-L55) I also see a key `3 digest_null "" cipher_null ""`), right?

viktor-kurchenko committed:
ci,ipsec: use cilium cli instead of bash scripts

The commit replaces bash scripts in CI with the Cilium CLI encrypt
sub-commands.

Signed-off-by: viktor-kurchenko <viktor.kurchenko@isovalent.com>

viktor-kurchenko committed:
doc,ipsec: update ipsec key create doc

Signed-off-by: viktor-kurchenko <viktor.kurchenko@isovalent.com>

viktor-kurchenko committed:
ipsec: fix expected key calculation

The commit fixes IPsec expected key calculation algorithm.

Signed-off-by: viktor-kurchenko <viktor.kurchenko@isovalent.com>

asauber reviewed (approved):
Looks good for CLI. GCM is a good default!

smagnani96 reviewed (approved):
LGTM, I also like simplifications in CI workflows to make full use of the `cilium encryption` command.

thorn3r reviewed (approved):
changes LGTM for sig-scalability, nice work :+1:

pchaigno commented on cilium-cli/cli/encrypt.go:
```diff
@@ -95,6 +96,29 @@ func newCmdIPsecKeyStatus() *cobra.Command {
+			return nil
+		},
+	}
+	cmd.Flags().StringVarP(&params.IPsecKeyAuthAlgo, "auth-algo", "", "gcm-aes", "IPsec key authentication algorithm (optional parameter, if omitted the current settings will be used). One of: gcm-aes, hmac-sha256, hmac-sha512")
```
Hm, how are the users supposed to know what "the current settings" are?

viktor-kurchenko replied:
Cobra will print the default value for the parameter:
```
cilium encrypt create-key --help
This command creates IPsec encryption key for the cluster

Usage:
  cilium encryption create-key [flags]

Flags:
      --auth-algo string         IPsec key authentication algorithm (optional parameter, if omitted the current settings will be used). One of: gcm-aes, hmac-sha256, hmac-sha512 (default "gcm-aes")
```

Isn't enough?

pchaigno replied:
Then I'd remove the text between parentheses.

viktor-kurchenko replied:
Done.

pchaigno commented on cilium-cli/cli/encrypt.go:
```diff
@@ -95,6 +96,29 @@ func newCmdIPsecKeyStatus() *cobra.Command {
+			return nil
+		},
+	}
+	cmd.Flags().StringVarP(&params.IPsecKeyAuthAlgo, "auth-algo", "", "gcm-aes", "IPsec key authentication algorithm (optional parameter, if omitted the current settings will be used). One of: gcm-aes, hmac-sha256, hmac-sha512")
```
The list of choices here is a bit confusing as it mixes AEAD algorithms with authentication algorithms and doesn't say anything of encryption algorithms.

viktor-kurchenko replied:
I see your point... What if I change the names to: `gcm-aes, hmac-sha256-aes, hmac-sha512-aes`?

pchaigno replied:
I think I'd go `rfc4106-gcm-aes`, `cbc-aes-sha256`, and `cbc-aes-sha512`. We could skip the RFC4106, but then we need to explain which variant of GCM AES we mean.

viktor-kurchenko replied:
Cool, let's do it!

viktor-kurchenko replied:
Done.

pchaigno commented on cilium-cli/encrypt/ipsec_new_key.go:
```diff
@@ -0,0 +1,67 @@
+func (s *Encrypt) checkEncryptionSecretNotExists(ctx context.Context) error {
+	_, err := s.client.GetSecret(ctx, s.params.CiliumNamespace, defaults.EncryptionSecretName, metav1.GetOptions{})
+	if err == nil {
+		return errors.New("IPsec secret already exists, please rotate key if needed")
```
Maybe let's point to the key rotation command as that may be what a user in this case is actually trying to do?

viktor-kurchenko replied:
Good idea, added. Thanks!

pchaigno reviewed (commented):
Looks good to me, though I think we should rework a bit the naming of the different algorithm options to avoid confusion.

liyihuang reviewed (approved):
LGTM on AWS CI

youngnick reviewed (approved):
LGTM on sig-servicemesh.

qmonnet commented on cilium-cli/encrypt/ipsec_new_key.go:
```diff
@@ -0,0 +1,67 @@
+func (s *Encrypt) checkEncryptionSecretNotExists(ctx context.Context) error {
+	_, err := s.client.GetSecret(ctx, s.params.CiliumNamespace, defaults.EncryptionSecretName, metav1.GetOptions{})
+	if err == nil {
+		return errors.New("IPsec secret already exists, please rotate key if needed using `cilium encryption rotate-key` command")
```
Nit: I'd remove “please”

```suggestion
		return errors.New("IPsec secret already exists, rotate key if needed using `cilium encryption rotate-key` command")
```

viktor-kurchenko replied:
Removed, thank you!

qmonnet reviewed (approved):
Docs look good, thank you!

viktor-kurchenko committed:
cli,ipsec: algorithm names improved

The commit improves IPsec algorithm names.

Signed-off-by: viktor-kurchenko <viktor.kurchenko@isovalent.com>

tommyp1ckles commented on cilium-cli/encrypt/ipsec_new_key.go:
```diff
@@ -0,0 +1,67 @@
+		return err
+	}
+
+	_, err = fmt.Printf("IPsec key successfully created, new key SPI: %d\n", newKey.spi)
```
Do we need to check the error of printf, think CI/unit tests will fail on bad formatting?

viktor-kurchenko replied:
I don't have a strong opinion here, does it better to drop the error and return nil:
```
	fmt.Printf("IPsec key successfully created, new key SPI: %d\n", newKey.spi)
	return nil
```
?

joestringer replied:
As a general rule it is bad practice to ignore errors.

But of course every rule has an exception :facepalm: ; The Go fmt package upstream documentation states that it is idiomatic to ignore errors specifically from `fmt.Printf()`. https://pkg.go.dev/fmt#Printf

joestringer replied:
Logically here the action (create the key) has actually been performed successfully. If there's some issue with getting this message to stdout then that is not consequential to executing the functionality requested by the user. That suggests that the error from `fmt.Printf()` should _not_ be returned. Again this is something I would typically not recommend but in this specific case it seems like the right pattern (to ignore the error from `fmt.Printf()`).

joestringer replied:
I guess to trace the full answer you'd want to look into the `Write()` implementation for `os.Stdout` https://cs.opensource.google/go/go/+/refs/tags/go1.24.3:src/os/file.go;l=191 ; at a glance the failure conditions are things like if this binary was outputting to stdout and that output was piped to another program which prematurely broke the pipe before reading all the input. I didn't dig much further than that, but I think that gives some reassurance that ignoring the error from fmt.Printf is OK here.

(Incidentally in a traditional unix command you would just silently return a zero exit code to indicate that the command executed successfully; printing like this would not be required. However printing a message like this can be useful if you're wanting to return information back to the user or just provide the user with more confidence that the command was completed successfully).

viktor-kurchenko replied:
@joestringer thank you for such detailed explanation!
I've removed the error check.

tommyp1ckles commented on cilium-cli/encrypt/ipsec_new_key.go:
```diff
@@ -0,0 +1,67 @@
+	if err == nil {
+		return errors.New("IPsec secret already exists, rotate key if needed using `cilium encryption rotate-key` command")
+	}
+	if err.Error() != fmt.Sprintf(`secrets "%s" not found`, defaults.EncryptionSecretName) {
```
This might prove to be brittle, if the error type for `secrets ... not found` is exported anywhere you could use `errors.Is(...)` (if this is coming from `client-go` then there is also a helper func for this: https://pkg.go.dev/k8s.io/apimachinery/pkg/api/errors#IsNotFound).

viktor-kurchenko replied:
Fixed, thank you!

tommyp1ckles reviewed (approved):
Looks, had a few points but they're minor. Approving to not hold this up anymore.
